### ISSUE TITLE
Enforce Production-Correct Training Semantics (DDP Accumulation, Scheduler Resume, bf16 Defaults, Dry-Run Boundary)

### ISSUE THREAD

**Issue by Anamitra-Sarkar** [Anamitra-Sarkar/llm_training_pipeline#3]
📌 CONTEXT & INTENT (READ CAREFULLY)

This repository is intended for real LLM pretraining / fine-tuning at scale, including:

Multi-GPU (DDP)

Long-running jobs

Frequent resume from checkpoints

bf16 mixed precision on modern accelerators

The current training script is structurally correct, but there are subtle correctness and contract issues that must be fixed explicitly to prevent silent training errors or Copilot “safe default” compromises.

This issue consolidates all required semantic guarantees into one authoritative contract.

🚨 NON-NEGOTIABLE ENGINEERING CONTRACT

Copilot must NOT:

downgrade defaults for safety

leave behavior implicit

rely on “works in practice”

skip edge cases because they don’t crash

Copilot must implement exactly the behaviors described below.

🔴 REQUIREMENT 1 — Correct Gradient Accumulation (DDP-Safe)
Problem

The current training loop performs gradient accumulation but does not scale the loss before backward.

This is mathematically incorrect under DDP and leads to:

incorrect gradient magnitudes

LR instability

silent divergence at scale

Mandatory Fix

Loss must be divided by gradient_accumulation_steps before backward():
loss = loss / config["gradient_accumulation_steps"]
loss.backward()

This must apply to:

single GPU

multi-GPU (DDP)

bf16 and fp32 modes

❌ Post-hoc averaging during logging is not sufficient.

🔴 REQUIREMENT 2 — Resume-Safe LR Scheduler Behavior
Problem

When resuming from a checkpoint:

model and optimizer are restored

LR scheduler continuity is not guaranteed

This causes:

LR jumps

warmup replay

silent optimization corruption

Mandatory Fix (choose ONE, but implement fully)

Option A (preferred):

Restore scheduler state directly from checkpoint

Option B (acceptable):

Explicitly fast-forward scheduler to start_step

Acceptance Rule

After resume:
LR(step = N + 1) == LR(step = N before interruption)

🔴 REQUIREMENT 3 — Production Defaults MUST Be Active

This repository is not a demo.

The following must be ENABLED BY DEFAULT, not optional conveniences:

bf16 mixed precision

AdamW optimizer

Learning-rate scheduler

Gradient checkpointing

Evaluation loop

Periodic checkpoint saving

These features may be disabled only via explicit opt-out flags (e.g., --no_bf16).

Copilot must not:

default to fp32 for safety

leave bf16 “available but unused”

gate production features behind examples

🔴 REQUIREMENT 4 — bf16 Autocast Semantics

bf16 handling must:

be enabled by default on supported CUDA GPUs

degrade gracefully on unsupported devices

not silently change precision modes mid-run

Preferred pattern:
with torch.cuda.amp.autocast(enabled=use_bf16):
    ...

Manual device branching should not introduce divergent behavior.

🔴 REQUIREMENT 5 — Dry-Run Execution Boundary (STRICT)

The purpose of --dry_run is verification, not training.

Dry-Run MUST:

load model and tokenizer

initialize optimizer and scheduler

load dataset and build dataloaders

run exactly one forward pass

exit cleanly

Dry-Run MUST NOT:

perform backward passes

update optimizer

step scheduler

enter training loops

perform evaluation loops

Dry-run exists to catch:

dependency errors

API mismatches

config wiring bugs

Nothing else.

🔴 REQUIREMENT 6 — No Implicit Safety Downgrades

Copilot must not:

“play safe” by weakening defaults

remove features to avoid edge cases

comment out behavior instead of fixing it

If something is complex:
➡️ Fix it, don’t disable it.

✅ ACCEPTANCE CRITERIA (ALL MUST PASS)

This issue is complete only when:

Gradient accumulation is mathematically correct under DDP

Resume preserves LR continuity exactly

bf16, checkpointing, eval are ON by default

Dry-run verifies initialization without training

No behavior regressions are introduced

Code remains readable and production-grade

🧠 ENGINEERING PHILOSOPHY (AUTHORITATIVE)

This repository targets:

long-running jobs

real hardware

real failures

Silent correctness bugs are worse than crashes.

Implementations must favor:

explicitness over cleverness

correctness over convenience

production semantics over demo safety

🔥 FINAL WARNING (INTENTIONAL)

If any of the above requirements are partially implemented, implicitly handled, or “assumed correct”, the issue is NOT resolved.

This is not an optimization request.
This is a semantic correctness contract.
